src/meili/match_route.cc

```diff
@@ -129,38 +134,47 @@ bool MergeRoute(const State& source,
 void cut_segments(const std::vector<MatchResult>& match_results,
                   int first_idx,
                   int last_idx,
-                  const std::vector<EdgeSegment>& segments,
+                  std::vector<EdgeSegment>& segments,
                   std::vector<EdgeSegment>& new_segments) {
   auto first_segment = segments.begin();
+  auto search_segment = first_segment;
   int prev_idx = first_idx;
   for (int curr_idx = first_idx + 1; curr_idx <= last_idx; ++curr_idx) {
     const MatchResult& curr_match = match_results[curr_idx];
     const MatchResult& prev_match = match_results[prev_idx];
 
-    // skip if we do not need to cut
-    if (!curr_match.is_break_point && curr_idx != last_idx) {
-      continue;
-    }
-
-    // Allow for some fp-fuzz in this gt comparison
+    // does the previous match point look further along its edge than the current
     bool prev_gt_curr = prev_match.distance_along > curr_match.distance_along + 1e-3;
 
-    // we want to handle to loop by locating the correct target edge by comparing the distance alone
+    // if they are on the same edge and previous comes after then its a loop
     bool loop = (prev_match.edgeid == curr_match.edgeid) && prev_gt_curr;
 
     // if it is a loop, we start the search after the first edge
-    auto last_segment = std::find_if(first_segment + static_cast<size_t>(loop), segments.end(),
-                                     [&curr_match](const EdgeSegment& segment) {
-                                       return (segment.edgeid == curr_match.edgeid);
-                                     });
+    for (search_segment += static_cast<size_t>(loop); search_segment != segments.end();
+         ++search_segment) {
+      if (search_segment->edgeid == curr_match.edgeid)
+        break;
+    }
 
-    if (last_segment == segments.cend()) {
+    if (search_segment == segments.cend()) {
       throw std::logic_error("In meili::cutsegments(), unexpectedly unable to locate target edge.");
     }
 
+    // if its not a break point nor the last segment, then we dont need to split it into two
+    if (!curr_match.is_break_point && curr_idx != last_idx) {
```

Comment:
so the main trick is we moved this continue down past where we did the loop to find the segment upon which the current match point lies. this way we have access to it, even if we arent going to cut it below. because i keep the `search_segment` around for each iteration of the main loop we are gauranteed that the inner loop does no more iterations than it otherwise would have and so this change shouldnt really cost us anything